src/broadcast.jl

```diff
@@ -8,199 +9,92 @@ else
-
-            func(B, As...)
-            return B
-        end
-    end  # let cache
+    ziptype(A) = Tuple{eltype(A)}
+    ziptype(A, B) = Zip2{Tuple{eltype(A)}, Tuple{eltype(B)}}
+    @inline ziptype(A, B, C, D...) = Zip{Tuple{eltype(A)}, ziptype(B, C, D...)}
 else
-    using Base.Broadcast: newindexer, map_newindexer, newindex
-
-    function _nullcheck(nargs)
-        nullcheck = :(isnull_1[I_1])
-        for i in 2:nargs
-            sym_isnull = Symbol("isnull_$i")
-            sym_idx = Symbol("I_$i")
-            nullcheck = Expr(:||, :($sym_isnull[$sym_idx]), nullcheck)
-        end
-        # if 0 argument arrays, treat nullcheck as though it returns false
-        nargs >= 1 ? nullcheck : :(false)
-    end
-
-    @generated function Base.Broadcast._broadcast!{K,ID,XT,nargs}(f,
-    Z::NullableArray, keeps::K, Idefaults::ID, Xs::XT, ::Type{Val{nargs}}; lift=false)
-        nullcheck = _nullcheck(nargs)
-        quote
-            T = eltype(Z)
-            $(Expr(:meta, :noinline))
-            # destructure keeps and Xs tuples (common to both lifted and non-lifted broadcast)
-            @nexprs $nargs i->(keep_i = keeps[i])
-            @nexprs $nargs i->(Idefault_i = Idefaults[i])
-            if !lift
-                # destructure the keeps and As tuples
-                @nexprs $nargs i->(X_i = Xs[i])
-                @simd for I in CartesianRange(indices(Z))
-                    # reverse-broadcast the indices
-                    @nexprs $nargs i->(I_i = newindex(I, keep_i, Idefault_i))
-                    # extract array values
-                    @nexprs $nargs i->(@inbounds val_i = X_i[I_i])
-                    # call the function and store the result
-                    @inbounds Z[I] = @ncall $nargs f val
-                end
-            else
-                # destructure the indexmaps and Xs tuples
-                @nexprs $nargs i->(values_i = Xs[i].values)
-                @nexprs $nargs i->(isnull_i = Xs[i].isnull)
-                @simd for I in CartesianRange(indices(Z))
-                    # reverse-broadcast the indices
-                    @nexprs $nargs i->(I_i = newindex(I, keep_i, Idefault_i))
-                    if $nullcheck
-                        # if any args are null, store null
-                        @inbounds Z.isnull[I] = true
-                    else
-                        # extract array values
-                        @nexprs $nargs i->(@inbounds val_i = values_i[I_i])
-                        # call the function and store the result
-                        @inbounds Z[I] = @ncall $nargs f val
-                    end
-                end
-            end
-        end
-    end
+    using Base.Broadcast: ziptype
+end
 
-    """
-        broadcast!(f, B::NullableArray, As::NullableArray...; lift::Bool=false)
+invoke_broadcast!{F, N}(f::F, dest, As::Vararg{NullableArray, N}) =
+    invoke(broadcast!, Tuple{Function, AbstractArray, Vararg{AbstractArray, N}}, f, dest, As...)
```

Comment:
Probably, I guess it's more correct if e.g. `f` is a type and there's a specific method for them.